frontend/src/metabase/visualizations/hoc/WithVizSettingsData.js

```diff
@@ -36,66 +39,70 @@ const WithVizSettingsData = ComposedComponent => {
       .filter(hasLinkedQuestionOrDashboard)
       .map(mapLinkedEntityToEntityQuery);
   }
-  return connect(
-    (state, props) => ({
-      getExtraDataForClick: clicked => {
-        const entitiesByTypeAndId = _.chain(getLinkTargets(clicked.settings))
-          .groupBy(target => target.entity.name)
-          .mapObject(targets =>
-            _.chain(targets)
-              .map(({ entity, entityId }) =>
-                entity.selectors.getObject(state, { entityId }),
-              )
-              .filter(object => object != null)
-              .indexBy(object => object.id)
-              .value(),
-          )
-          .value();
-        return {
-          ...entitiesByTypeAndId,
-          parameterValuesBySlug: props.parameterValuesBySlug,
-          dashboard: props.dashboard,
-          dashcard: props.dashcard,
-          userAttributes: getUserAttributes(state, props),
-        };
-      },
-    }),
-    dispatch => ({ dispatch }),
-  )(
-    class WithVizSettingsData extends React.Component {
-      dashcardSettings(props) {
-        const [firstSeries] = props.rawSeries || [{}];
-        const { visualization_settings } = firstSeries.card || {};
-        return visualization_settings;
-      }
-
-      fetch() {
-        getLinkTargets(this.dashcardSettings(this.props)).forEach(
-          ({ entity, entityId }) =>
-            this.props.dispatch(
-              entity.actions.fetch({ id: entityId }, { noEvent: true }),
-            ),
-        );
-      }
+  return withRouter(
+    connect(
+      (state, props) => ({
+        getExtraDataForClick: clicked => {
+          const entitiesByTypeAndId = _.chain(getLinkTargets(clicked.settings))
+            .groupBy(target => target.entity.name)
+            .mapObject(targets =>
+              _.chain(targets)
+                .map(({ entity, entityId }) =>
+                  entity.selectors.getObject(state, { entityId }),
+                )
+                .filter(object => object != null)
+                .indexBy(object => object.id)
+                .value(),
+            )
+            .value();
+          return {
+            ...entitiesByTypeAndId,
+            location: props.location,
+            routerParams: props.params,
+            parameterValuesBySlug: props.parameterValuesBySlug,
+            dashboard: props.dashboard,
+            dashcard: props.dashcard,
+            userAttributes: getUserAttributes(state, props),
+          };
+        },
+      }),
```

Comment:
The diff highlight is going crazy here. There are only two changes here:

1. We're now adding router `location` and `params` props to the final object here
2. The whole thing is now wrapped into `withRouter` HoC to ensure these props are available